### PR TITLE
[Better Engineering] Bump ruff to 0.0.278 and fix new lint errors

### PR DESCRIPTION
**Pull request by justinchuby**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #16789

Bump ruff to 0.0.278 and fix new lint errors. I added noqa to all existing RUF012 errors which requires mutable class variables to be annotated with `ClassVar`, as well as all PERF issues.

Signed-off-by: Justin Chu <justinchu@microsoft.com>

### CODE REVIEW

File: onnxruntime/test/python/onnxruntime_test_training_unit_tests.py [hunk 1]
```diff
@@ -12,18 +12,17 @@
...
-        # to sort backward graph before forward graph which gives incorrect result.
-        # TODO Re-enable when that is fixed.
-        return
+    def setUp(self):
+        torch.manual_seed(1)
+        onnxruntime.set_seed(1)
 
+    @unittest.skip(
+        "Temporarily disable this test. The graph below will trigger ORT to "
+        "sort backward graph before forward graph which gives incorrect result."
```
Comment (jcwchen) [inline onnxruntime/test/python/onnxruntime_test_training_unit_tests.py]: Remain TODO for more obvious reminder?

File: orttraining/orttraining/python/checkpointing_utils.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ def _split_name(self, name):
         elif name_split[0].endswith("_fp16"):
             mp_suffix = "_fp16"
         param_name = name_split[0]
-        if optimizer_key != "":  # noqa: PLC1901
```
Comment (jcwchen) [inline orttraining/orttraining/python/checkpointing_utils.py:72]: Just for my learning: why we don't need this anymore without any update? Were they (PLC1901 errors) false positive or they are not required for the latest ruff anymore?

**Review by jcwchen — DISMISSED** [review]
nit comments.

File: pyproject.toml [hunk 3]
```diff
@@ -45,20 +45,22 @@ reportMissingImports = false
 # NOTE: Do not create an exclude list. Edit .lintrunner.toml instead
 target-version = "py38"
 select = [
+    "B", # flake8-bugbear
```
Comment (justinchuby) [inline pyproject.toml:48]: Sorted

**Review by jcwchen — APPROVED** [review]

**Pull request merged**